File: code/src/main/scala/basket/apriori.scala
package basket

/*
 name:
 netlink id:
 */

object aPriori {

  /**
   *  fold the newItems into the existing items
   *    - if the key for the new item is present, add the count of the items, else
   *    - default to the value of the new item
   * @param items the current set of items
   * @param newItems the new items to merge into the map
   * @tparam K key type
   * @return Merged map
   */
  def mergeAndSumMaps[K](items: Map[K, Int], newItems:Map[K,Int]) : Map[K,Int] =
  {
    newItems.foldLeft(items) {
      case (item_set, (key,value)) => item_set + (key -> item_set.get(key).map(_ + value).getOrElse(value))
    }
  }

  /**
   * Performs first pass of the apriori algorithm
   *
   * @param threshold Threshold of item frequency to keep (based on total count)
   * @param lines Iterator of open file. FILE IS LARGE
   * @param delim expected delimiter to split basket lines on
   * @return 3 tuple (count of items in basket:Int, support:Int, frequent items: Items)
   */
  def doFirstPass(threshold: Double, lines: Iterator[String], delim:String): (Int, Int, Items) =
  {
    val (count, items):(Int,Items) = lines.foldLeft((0, Map[Elem,Int]()))( (accum, line) => {
      val newItemMap: Items = extractItemsFromLine(line, delim)
      (accum._1 + 1, mergeAndSumMaps(accum._2, newItemMap))
    })

    // calculate support and filter based on this
    val support:Int = (count * threshold).toInt
    val filtered:Items = items.filter(_._2 >= support)

    (count, support, filtered)
  }

  def extractItemsFromLine(line: String, delim: String): Items =
  {
    val itemsInBasket:List[String] = line.split(delim).toList
    val cleaned:List[String] = itemsInBasket.map(_.trim())
    cleaned.zip(List.fill(itemsInBasket.length)(1)).toMap
  }

  /**
   * Performs second pass of the apriori algorithm
   *
   * @param supportT threshold of item frequency to keep
   * @param items frequent items determined in first pass of apriori algorithm
   * @param lines Iterator of open file. FILE IS LARGE
   * @param delim expected delimiter to split basket lines on
   * @return Frequent pairs (BasketItem1,BasketItem2)->Frequency. There is only one (a,b) entry, i.e if (a,b) exists, (b,a) will not
   */
  def doSecondPass(supportT: Int, items: Items, lines: Iterator[String], delim:String): FreqPairs =
  {
    val freqPairs:FreqPairs = lines.foldLeft(Map[(Elem,Elem), Int]())((accum, line) => {
      val basketItemMap: Items = extractItemsFromLine(line, delim)

      val frequentItems: Iterable[Elem] = for {
        item <- basketItemMap
        if (items.contains(item._1))
      } yield item._1

      // first generate all tuples
      val freqPairsT = (for {
        f1 <- frequentItems
        f2 <- frequentItems

        if (f1 != f2)
      } yield (f1,f2)->1).toMap

      // then remove the ones that are cases of (a,b), (b,a) within basket
      // i.e. if list was [(a,b)->1, (b,a)->1], this would return [(a,b)->1]
      val newFreqMap: FreqPairs = freqPairsT.foldLeft(Map[(Elem,Elem),Int]())(
        (op, pair) =>
        {
          if (op.contains((pair._1._2, pair._1._1))) op else op + pair
        }
      )

      // merge these results with the running results
      mergeAndSumMaps(accum, newFreqMap)
    }).foldLeft (Map[(Elem,Elem),Int]())(
      // If we found (a,b) in one basket and (b,a) in another basket, we need to handle this case by summing them and
      // keeping only one entry
      (op, pair) =>
      {
        val freq = pair._2 + op.get((pair._1._2, pair._1._1)).getOrElse(0)
        if (op.contains((pair._1._2, pair._1._1))) op + ((pair._1._2, pair._1._1) -> freq) else op + (pair._1 -> freq)
      }
    ). filter (_._2 >= supportT)

    freqPairs
  }

  /**
   * Get statistics for a given pair
   *
   * @param count count of total baskets
   * @param items single item frequency
   * @param pair pair of elements (statistics will be generated for (1->2)
   * @param pairFreq frequency of the pair
   * @return Option[Result] maintaining the statistics, or None if input is malformed
   */
  def getPairStats(count:Int, items:Items, pair:(Elem,Elem), pairFreq: Int) : Option[Result] =
  {
    (items.get(pair._1), items.get(pair._2)) match {
      case (Some(itemFreq), Some(pairItemFreq)) =>
        // Confidence {1} -> {2} = count(1,2)/count(1)
        val confidence: Double = pairFreq.toDouble / pairItemFreq

        val supportPair: Double = pairFreq.toDouble / count
        val supportItem: Double = itemFreq.toDouble / count

        // lift {1}->{2} = (confidence(1,2) / (fraction_of_items_containing(1))
        // val lift: Double = (confidence) / (itemFreq.toDouble / count)
        // different formula, supportPair / (itemSupport * otherItemSupport)
        val lift = supportPair / (supportItem * (pairItemFreq.toDouble / count))

        //    val interest = confidence - (items.get(pair._2) / count)
        val itemName = pair._1
        val boughtWith = pair._2

        Some(
          Result(
            item = itemName,
            freqItem = itemFreq,
            support = precision(supportItem),
            boughtWith = boughtWith,
            freqPair = pairFreq,
            supportPair = precision(supportPair),
            confidence = precision(confidence),
            lift = precision(lift))
        )
      case _ => None
    }
  }

  def precision(v: Double):Double = {
    val digits = 7
    Math.floor(v * Math.pow(10,digits))/Math.pow(10, digits)
  }

  /**
   * Get results for pair frequency and print results
   *
   * @param count count of items in basket
   * @param items individual item frequency
   * @param freqPairs pair frequency
   * @param limit max number of records to print, or None
   */
  def doResults(count:Int, items: Items, freqPairs: FreqPairs, limit: Option[Int]) =
  {
    val interimResults: List[Option[Result]] = freqPairs.foldLeft(List[Option[Result]]())((op, freqPair) => {
      val result1: Option[Result] = getPairStats(count, items, freqPair._1, freqPair._2)
      val result2: Option[Result] = getPairStats(count, items, (freqPair._1._2, freqPair._1._1), freqPair._2)
      op ++ List[Option[Result]](result1, result2)
    })

    // first convert from List[Option[Result]] to List[Result] and then sort
    // there should be no "None" here but including for completion
    val resultsSorted: List[Result] = interimResults.foldLeft(List[Result]())((op, res) => {
      res match
      {
        case Some(r) => op :+ r
        case None => op
      }
    }).sortBy(r => (-r.confidence, -r.lift, -r.freqItem, r.item, r.boughtWith))

    // limit if necessary
    val (resultsFiltered, realLimit) = limit match {
      case Some(realLimit) => (resultsSorted.take(realLimit), realLimit)
      case _ => (resultsSorted, 0)
    }

    output.printResults(resultsFiltered)
    println(s"Found ${resultsSorted.length} records." +
      { if (realLimit != 0 && realLimit < resultsSorted.length) s" Printed only $realLimit." else "" } )
  }
}

File: python_impl/apriori.py
from argparse import ArgumentError
from dataclasses import dataclass
import math
import sys
from typing import Tuple
from decimal import Decimal

@dataclass
class Result():
    item: str
    boughtWith: str
    freqItem: int
    freqPair: int
    support: float
    supportPair: float
    confidence: float
    lift: float

def precision(v: float) -> float:
    """
    apriori.scala used 7 digits for truncation, but python does some weird rounding stuff so I needed to use 8 digits here
    """
    digits = 8
    return float(math.floor(v * 10**digits)/(10**digits))

class APriori:
    def __init__(self):
        """ """

    def firstPass(self, filename:str, delim:str, threshold:float) -> Tuple[int, int, dict[str, int]]:
        """
        perform first pass of the apriori algorithm

        inputs: filename - file to iterate
        delim: delimiter to separate basket items
        threshold: number between (0,1] to determine support threshold for items

        output: (basket count, dict[item,frequency])
        """
        freq_items_unfiltered = {}
        count = 0

        with open(filename, "r") as in_file:
            for line in in_file:
                basket = set([w.strip() for w in line.split(delim)])
                for item in basket:
                    if item in freq_items_unfiltered:
                        freq_items_unfiltered[item] += 1
                    else:
                        freq_items_unfiltered[item] = 1

                count += 1

        support = int(count * threshold)
        freq_items_filtered = {}

        for key in freq_items_unfiltered:
            if freq_items_unfiltered[key] >= support:
                freq_items_filtered[key] = freq_items_unfiltered[key]

        return (count, support, freq_items_filtered)


    def permute(self, lst: list):
        """
        returns all permutations of a list as tuples
        """
        if len(lst) > 1:
            for idx, item in enumerate(lst):
                for item2 in lst[idx+1:]:
                    yield (item,item2)
        else:
            return []

    def secondPass(self, filename:str, delim:str, support:int, items:dict[str,int]) -> dict[(str,str), int]:
        """
        perform second pass of the apriori algorithm

        inputs: filename - file to iterate
        delim: delimiter to separate basket items
        support: minimum frequency to keep

        output: (dict[(frequent item1, bought with item),frequency]) about frequency specified by `support`
        """

        freq_pairs = {}

        with open(filename, "r") as in_file:
            for line in in_file:
                frequent = []
                basket = set([w.strip() for w in line.split(delim)])
                [frequent.append(item) for item in basket if item in items]

                permutations = list(self.permute(frequent))

                for permutation in permutations:
                    if permutation in freq_pairs:
                        freq_pairs[permutation] += 1
                    elif (permutation[1], permutation[0]) in freq_pairs:
                        freq_pairs[(permutation[1], permutation[0])] += 1
                    else:
                        freq_pairs[permutation] = 1

        freq_pairs_filtered = {}

        for key in freq_pairs:
            if (key[1], key[0]) in freq_pairs:
                print(f"Key {key} had a reverse mapping")

        for key in freq_pairs:
            if freq_pairs[key] >= support:
                freq_pairs_filtered[key] = freq_pairs[key]

        return freq_pairs_filtered

    def _getResult(self, count: int, items: dict[str, int], pair: Tuple[str, str], pairFreq: int) -> Result:
        """
        Calculate statistics of an item pair 
        """
        itemFreq = items[pair[0]]
        pairItemFreq = items[pair[1]]

        confidence = float(pairFreq) / pairItemFreq

        supportPair = float(pairFreq) / count
        supportItem = float(itemFreq) / count

        # val lift = supportPair / (supportItem * (pairItemFreq.toDouble / count))
        # lift = confidence / (float(itemFreq) / count)
        lift = float(supportPair) / (supportItem * (float(pairItemFreq) / count))

        itemName = pair[0]
        boughtWith = pair[1]

        return Result(itemName, boughtWith, itemFreq, pairFreq, precision(supportItem), precision(supportPair), precision(confidence), precision(lift))


    def getResults(self, count: int, items: dict[str, int], freq_pairs: dict[Tuple[str,str], int], limit=None):
        """ 
        Get results from frequency pairs and print them to screen
        
        """

        results = []
        for pair in freq_pairs:
            res1 = self._getResult(count, items, pair, freq_pairs[pair])
            res2 = self._getResult(count, items, (pair[1], pair[0]), freq_pairs[pair])

            results += [res1, res2]

        sorted_list = sorted(results, key=lambda res: (-res.confidence, -res.lift, -res.freqItem, res.item, res.boughtWith))

        if limit:
            self.printResults(sorted_list[:limit])
        else:
            self.printResults(sorted_list)

        print("Found {0} records.{1}".format(len(sorted_list), " Printed only {0}.".format(limit) if limit and limit < len(results) else ""))

    def printResults(self, results: list[Result]):
        maxWidth = max(20, len(max(results, key=lambda res:len(res.item)).item) + 2)

        header = "SupportPair-{item:<{maxWidth}}- Freq -Support-{boughtWith:{maxWidth2}}- FreqPair-Confidence-     Lift".format(item="Item", maxWidth=maxWidth, boughtWith="Bought with", maxWidth2=maxWidth-4)
        sep = ['-'] * len(header)
        print(header)
        print(''.join(sep))
        # s"  %9.6f %-${maxWidth}s %5d %8.6f %-${maxWidth2}s %5d %10.2f %9.3f".format(
        for res in results:
            print("  {0:9.6f} {1:{maxWidth}} {2:5d} {3:8.6f} {4:{maxWidth2}} {5:5d} {6:10.2f} {7:9.3f}".format(
                res.supportPair, res.item, res.freqItem, res.support, res.boughtWith, res.freqPair, 
                res.confidence, res.lift, maxWidth=maxWidth, maxWidth2=maxWidth
            ))
        print(''.join(sep))

if __name__ == "__main__":
    if len(sys.argv) < 4 or len(sys.argv) > 5:
        raise Exception("Arguments missing <filename> <delimiter> <relativeSupportThreshold> [limit] ")

    filename:str = sys.argv[1]
    delim:str = sys.argv[2]

    if (len(delim) != 1):
        raise ArgumentError(f"Delimiter should be one character ([{delim}] provided)")

    threshold = float(sys.argv[3])

    limit = int(sys.argv[4]) if len(sys.argv) == 5 else None

    print("Running with parameters: Filename [{0}] Separator [{1}] Minimum relative support threshold [{2}].{3}".format(
        filename, delim, threshold,
        "" if limit is None else f" Print at most {limit} tuples."
    ), file=sys.stderr)
    # print("" if limit is None else f"Print at most {limit} tuples.", file=sys.stderr)
    # print("")

    apriori = APriori()
    (count, support, items) = apriori.firstPass(filename, delim, threshold)

    print(f"""{count} records, only {len(items)} item{"s" if (len(items) > 1) else ""} above support threshold {support} ({threshold}).\n""")

    freq_pairs = apriori.secondPass(filename, delim, support, items)

    # print(f"""{len(freq_pairs)} frequent pair{"s" if (len(freq_pairs) > 1) else ""}""")

    apriori.getResults(count, items, freq_pairs, limit)
